drivers/flash/soc_flash_silabs_series2.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2018, Piotr Mienkowski
+ * Copyright (c) 2025 Silicon Laboratories Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT     silabs_series2_flash_controller
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/dma.h>
+#include <soc.h>
+#include <em_msc.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(flash_silabs, CONFIG_FLASH_LOG_LEVEL);
+
+struct flash_silabs_data {
+	struct k_sem mutex;
+#if defined(CONFIG_SOC_FLASH_SILABS_S2_DMA_WRITE) || defined(CONFIG_SOC_FLASH_SILABS_S2_DMA_READ)
+	const struct device *dma_dev;
+	uint32_t dma_channel;
+#endif
+#if defined(CONFIG_SOC_FLASH_SILABS_S2_DMA_READ)
+	struct k_sem sync;
+	int sync_status;
+	struct dma_config dma_cfg;
+	struct dma_block_config dma_block_cfg;
+#endif
+};
+
+struct flash_silabs_config {
+	struct flash_parameters flash_parameters;
+	struct flash_pages_layout page_layout;
+	bool lpwrite;
+};
+
+static bool flash_silabs_read_range_is_valid(off_t offset, uint32_t size)
+{
+	return (offset + size) <= DT_REG_SIZE(SOC_NV_FLASH_NODE);
```

Comment:
`offset` may be negative, is seems as there is not check.
Also `(offset + size) <= DT_REG_SIZE(SOC_NV_FLASH_NODE)` may give true if `offset + size` does integer overflow below the right size value of comparison.
This should rather be
```
(offset >= 0) && (offset < DT_REG_SIZE(SOC_NV_FLASH_NODE) && ((DT_REG_SIZE(SOC_NV_FLASH_NODE) - offset) >= size)
```